genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -0,0 +1,363 @@
+                }
+            }
+        }
+    }
+
+    private synchronized void pushManifest() {
+        if (started.get()) {
+            final AgentManifestMessage jobFileManifest;
+            try {
+                jobFileManifest = manifestProtoConverter.manifestToProtoMessage(
+                    this.jobId,
+                    new JobDirectoryManifest(this.jobDirectoryPath, false)
+                );
+            } catch (final IOException e) {
+                log.error("Failed to construct manifest", e);
+                return;
+            } catch (GenieConversionException e) {
+                log.error("Failed to serialize manifest", e);
+                return;
+            }
+
+            if (this.controlStreamObserver == null) {
+                this.controlStreamObserver = fileStreamServiceStub.sync(this.responseObserver);
+                if (this.controlStreamObserver instanceof ClientCallStreamObserver) {
+                    ((ClientCallStreamObserver) this.controlStreamObserver).setMessageCompression(ENABLE_COMPRESSION);
+                }
+            }
+
+            this.controlStreamObserver.onNext(jobFileManifest);
+        }
+    }
+
+    private void handleControlStreamError(final Throwable t) {
+        log.warn("Manifest stream error: {}", t.getMessage(), t);
+        this.trigger.reset();
+        this.discardCurrentStream(false);
+    }
+
+    private void handleControlStreamCompletion() {
+        log.debug("Manifest stream completed");
+        this.discardCurrentStream(false);
+    }
+
+    private synchronized void discardCurrentStream(final boolean sendStreamCompletion) {
+        if (this.controlStreamObserver != null) {
+            if (sendStreamCompletion) {
+                this.controlStreamObserver.onCompleted();
+            }
+            this.controlStreamObserver = null;
+        }
+    }
+
+    private synchronized void handleFileRequest(
+        final String streamId,
+        final String relativePath,
+        final int startOffset,
+        final int endOffset
+    ) {
+        log.info(
+            "Server is requesting file {} (range: [{}, {}, streamId: {})",
+            relativePath,
+            startOffset,
+            endOffset,
+            streamId
+        );
+
+        if (!this.started.get()) {
+            log.warn("Ignoring file request, service shutting down");
+            return;
+        }
+
+        final Path absolutePath = this.jobDirectoryPath.resolve(relativePath);
+
+        if (!Files.exists(absolutePath)) {
+            log.warn("Ignoring request for a file that does not exist: {}", absolutePath);
+            return;
+        }
+
+        final boolean permitAcquired = this.concurrentTransfersSemaphore.tryAcquire();
+
+        System.out.println("<<< " + this.concurrentTransfersSemaphore.availablePermits());
+
+        if (!permitAcquired) {
+            log.warn("Ignoring file request, too many transfers already in progress");
```

Comment:
Shouldn't this propagate an error?

Reply:
All failures of the agent to send back the requested file are treated the same: timeout on the server and resource stream closed.
This is because even if we did send back a more specific error to the server, there's no way to expose it. The `Resource` interface specifies the kind of errors thrown and that's all we have.